File: api/src/main/java/com/wbrawner/twigs/user/UserRepository.java
package com.wbrawner.twigs.user;

import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.PagingAndSortingRepository;

import java.util.List;
import java.util.Optional;

public interface UserRepository extends CrudRepository<User, String>,
        PagingAndSortingRepository<User, String> {
    Optional<User> findByUsername(String username);

    Optional<User> findByUsernameAndPassword(String username, String password);

    List<User> findByUsernameContains(String username);

    Optional<User> findByEmail(String email);
}

File: api/src/main/kotlin/com/wbrawner/twigs/server/user/UserRepository.kt
package com.wbrawner.twigs.server.user

import org.springframework.data.repository.PagingAndSortingRepository
import java.util.*

interface UserRepository : PagingAndSortingRepository<User, String> {
    fun findByName(name: String?): Optional<User>
    fun findByNameContains(name: String?): List<User>
    fun findByEmail(email: String?): Optional<User>
}